csrc/scheduler/normalization_inner_outer.cpp

```diff
@@ -0,0 +1,857 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <ATen/cuda/CUDAContext.h>
+#include <inlining.h>
+#include <instrumentation.h>
+#include <scheduler/debug_utils.h>
+#include <scheduler/normalization_inner_outer.h>
+#include <scheduler/normalization_utils.h>
+#include <scheduler/reduction_utils.h>
+#include <scheduler/utils.h>
+
+namespace nvfuser {
+
+constexpr auto schedule_heuristic = ScheduleHeuristic::InnerOuterPersistent;
+
+std::shared_ptr<ReductionParams> getInnerOuterPersistentHeuristics(
+    Fusion* fusion,
+    const at::ArrayRef<c10::IValue>& runtime_inputs,
+    HeuristicSummary* data_cache) {
+  FUSER_PERF_SCOPE("getInnerOuterPersistentHeuristicsFromIValue");
+  SchedulerRuntimeInfo runtime_info(fusion, runtime_inputs);
+  auto reduction_type = reduction_scheduler_utils::getReductionType(fusion);
+  NVF_CHECK(
+      reduction_type == reduction_scheduler_utils::ReductionType::InnerOuter,
+      "Reduction type should be ReductionType::InnerOuter.");
+  return InnerOuterPersistentKernelScheduler::getPersistentHeuristic(
+      fusion, runtime_info, data_cache);
+}
+
+InnerOuterPersistentKernelScheduler::InnerOuterPersistentKernelScheduler(
+    Fusion* fusion,
+    SchedulerRuntimeInfo& runtime_info,
+    HeuristicSummary* data_cache)
+    : SchedulerEntry(schedule_heuristic) {
+  computeHeuristics(fusion, runtime_info, data_cache);
+}
+
+void InnerOuterPersistentKernelScheduler::computeHeuristics(
+    Fusion* fusion,
+    SchedulerRuntimeInfo& runtime_info,
+    HeuristicSummary* data_cache) {
+  params_ = getPersistentHeuristic(fusion, runtime_info, data_cache);
+  NVF_ERROR(params_ != nullptr);
+}
+
+void InnerOuterPersistentKernelScheduler::schedule(Fusion* fusion) {
+  FUSER_PERF_SCOPE("Schedule InnerOuterPersistent Fusion");
+  schedulePersistentKernel(fusion, reductionParams());
+}
+
+bool InnerOuterPersistentKernelScheduler::canScheduleCompileTime(
```

Comment:
adapted from the original compile time check. Steps (1), (2), and (4) are same for all persistent kernels, and they are abstracted into utility functions.